flake8_trio.py

```diff
@@ -255,6 +255,99 @@ def check_for_trio100(self, node: Union[ast.With, ast.AsyncWith]) -> None:
+        and isinstance(node.value, ast.Name)
+        and node.value.id == "trio"
+        and node.attr == "Cancelled"
+    )
+
+
+# Never have an except Cancelled or except BaseException block with a code path that
+# doesn't re-raise the error
+class Visitor103(ast.NodeVisitor):
+    def __init__(self) -> None:
+        super().__init__()
+        self.problems: List[Error] = []
+
+        # Likely overkill, but need to figure out nested try's first.
+        self.except_names: List[Optional[str]] = []
+        self.unraised: bool = False
+
+    def visit_ExceptHandler(self, node: ast.ExceptHandler):
+        outer_unraised = self.unraised
+        must_raise = False
+        if node.type is not None:
+            if (
+                isinstance(node.type, ast.Tuple)
+                and any(has_exception(v) for v in node.type.elts)
+            ) or has_exception(node.type):
+                must_raise = True
+                self.except_names.append(node.name)
+                self.unraised = True
+
+        self.generic_visit(node)
+        if must_raise:
+            if self.unraised:
+                # TODO: point at correct exception
+                self.problems.append(make_error(TRIO103, node.lineno, node.col_offset))
+            self.except_names.pop()
+
+        self.unraised = outer_unraised
+
+    def visit_Raise(self, node: ast.Raise):
+        if self.unraised:
+            assert self.except_names
+            name = self.except_names[-1]
+            if (name is None and node.exc is None) or (
+                isinstance(node.exc, ast.Name) and node.exc.id == name
+            ):
+                self.unraised = False
+            else:
+                # Error: something other than the exception was raised
+                # TODO: give separate error message
+                self.problems.append(make_error(TRIO103, node.lineno, node.col_offset))
+                self.unraised = False
+        self.generic_visit(node)
+
+    def visit_Return(self, node: ast.Return):
+        if self.unraised:
+            # Error, only good way of stopping the function is to raise
+            self.problems.append(make_error(TRIO103, node.lineno, node.col_offset))
+            self.unraised = False
+        self.generic_visit(node)
+
+    def visit_If(self, node: Union[ast.If, ast.For, ast.While]):
+        if not self.unraised:
+            self.generic_visit(node)
+            return
+
+        body_raised = False
+        for n in node.body:
+            self.visit(n)
+
+        # does body always raise correctly
+        body_raised = not self.unraised
+
+        self.unraised = True
+        for n in node.orelse:
+            self.visit(n)
+        # does orelse always raise correctly
+        orelse_raised = not self.unraised
+
+        # if both paths always raises, unset unraised.
+        self.unraised = not (body_raised and orelse_raised)
+
+    # checking for/while in this context doesn't make *that* much sense
```

Comment:
I find the theoretical case for this highly amusing:
```python
try:
  pass
except trio.Cancelled:
  while <...>:
    if <...>:
      break
    raise
  else:
    raise
```
is in theory totally acceptable - but it's just a very funky if statement.

... nowait nevermind me, brainfart. This is not how [while/for] / else works. Need some logic for break, or scream if somebody tries to do something stupid like this. Will fix

Reply:
Okay I think I'll just disregard all `raise`'s inside loops. I can't come up with any pattern where that is good code, and handling the bad code is a pain after all. But any return's in there need to be screamed about.

Reply:
thanks for coming to my TED talk